clippy_lints/src/use_crate_prefix_for_self_imports.rs

```diff
@@ -0,0 +1,149 @@
+    pub USE_CRATE_PREFIX_FOR_SELF_IMPORTS,
+    nursery,
+    "checks that imports from the current crate use the `crate::` prefix"
+}
+
+#[derive(Clone, Default)]
+pub struct UseCratePrefixForSelfImports<'a, 'tcx> {
+    /// collect `use` in current block
+    use_block: Vec<&'a UsePath<'tcx>>,
+    /// collect `mod` in current block
+    mod_names: FxHashSet<Symbol>,
+    /// spans of `mod`, `use`, and attributes
+    spans: Vec<Span>,
+}
+
+impl_lint_pass!(UseCratePrefixForSelfImports<'_, '_> => [USE_CRATE_PREFIX_FOR_SELF_IMPORTS]);
+
+impl<'a, 'tcx> LateLintPass<'tcx> for UseCratePrefixForSelfImports<'a, 'tcx> {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'a Item<'tcx>) {
+        let FileName::Real(RealFileName::LocalPath(p)) = cx.sess().source_map().span_to_filename(item.span) else {
+            self.clear();
+            return;
+        };
+        let Some(file_name) = p.file_name() else {
+            self.clear();
+            return;
+        };
+        // only check `main.rs` and `lib.rs`
+        if !(file_name == "main.rs" || file_name == "lib.rs") {
+            return;
+        }
+
+        if self.in_same_block(item.span) {
+            self.insert_item(item);
+        } else {
+            self.deal(cx);
+            self.clear();
+            self.insert_item(item);
+        }
+    }
+}
+
+impl<'tcx> UseCratePrefixForSelfImports<'_, 'tcx> {
+    fn in_same_block(&self, span: Span) -> bool {
+        if self.spans.is_empty() {
+            return true;
+        }
+        if self.spans.iter().any(|x| x.contains(span)) {
+            return true;
+        }
+        if self.spans.iter().any(|x| span.lo() - x.hi() == BytePos(1)) {
+            return true;
+        }
+        false
+    }
+
+    fn insert_item(&mut self, item: &Item<'tcx>) {
+        match item.kind {
+            ItemKind::Mod(ident, _) => {
+                self.spans.push(item.span);
+                self.mod_names.insert(ident.name);
+            },
+            ItemKind::Use(use_tree, _) => {
+                self.spans.push(item.span);
+                self.use_block.push(use_tree);
+            },
+            _ => {},
+        }
+    }
+
+    fn deal(&self, cx: &LateContext<'tcx>) {
+        for use_path in &self.use_block {
+            if let Some(segment) = use_path.segments.first()
+                && let Res::Def(_, def_id) = segment.res
+                && def_id.krate == LOCAL_CRATE
+            {
+                let root = segment.ident.name;
+                if root != rustc_span::symbol::kw::Crate
+                    && root != rustc_span::symbol::kw::Super
+                    && root != rustc_span::symbol::kw::SelfLower
+                    && !self.mod_names.contains(&root)
+                {
+                    span_lint_and_sugg(
+                        cx,
+                        USE_CRATE_PREFIX_FOR_SELF_IMPORTS,
+                        segment.ident.span,
+                        "this import is not clear",
```

Comment:
This *message* is not clear. Why is the import "not clear"? (I don't have a better suggestion yet)

Reply:
> Why is the import "not clear"?

I can't infer whether it's from crate or from my local codebase